File: 05-06-2025_Count_the_paths.java
class Solution {
    public int countPaths(int[][] edges, int V, int src, int dest) {

        // to make graph
        List<List<Integer>> graph = new ArrayList<>();
        int[] indegree = new int[V];
        for (int i = 0; i < V; i++) graph.add(new ArrayList<>());

        // add all edges and calculate in-degree
        for (int[] e : edges) {
            graph.get(e[0]).add(e[1]);
            indegree[e[1]]++;
        }

        // to make topological sorting
        Queue<Integer> q = new LinkedList<>();

        // insert all vertices which don't have any parent
        for (int i = 0; i < V; i++) {
            if (indegree[i] == 0) q.add(i);
        }

        List<Integer> topo = new ArrayList<>();

        // using kahn's algorithm for topological sorting
        while (!q.isEmpty()) {
            int node = q.poll();
            topo.add(node);
            for (int nei : graph.get(node)) {
                indegree[nei]--;
                if (indegree[nei] == 0) q.add(nei);
            }
        }

        // to store number of ways
        int[] dp = new int[V];
        dp[dest] = 1;

        // traverse in reverse topological order
        for (int i = topo.size() - 1; i >= 0; i--) {
            int u = topo.get(i);
            for (int nei : graph.get(u)) {
                dp[u] += dp[nei];
            }
        }

        return dp[src];
    }
}

File: 05-06-2025_Count_the_paths.py
class Solution:

    def countPaths(self, edges, V, src, dest):

        # to make graph
        graph = [[] for _ in range(V)]
        indegree = [0] * V

        # add edges and calculate in-degree
        for u, w in edges:
            graph[u].append(w)
            indegree[w] += 1

        # to make topological sorting
        q = deque()

        # insert all vertices which don't have any parent
        for i in range(V):
            if indegree[i] == 0:
                q.append(i)

        topo = []

        # using kahn's algorithm for topological sorting
        while q:
            node = q.popleft()
            topo.append(node)
            for nei in graph[node]:
                indegree[nei] -= 1
                if indegree[nei] == 0:
                    q.append(nei)

        # to store number of ways
        dp = [0] * V
        dp[dest] = 1

        # traverse in reverse topological order
        for node in reversed(topo):
            for nei in graph[node]:
                dp[node] += dp[nei]

        return dp[src]